### DISCUSSION TITLE
Updatable consumer settings

### DISCUSSION THREAD

**Discussion by scottf** [nats-io/nats-architecture-and-design#87] [General]
According to unit tests in the Java client, you can update these settings in a consumer:

- deliverSubject
- ackWait
- rateLimit
- maxAckPending
- maxDeliver

and these settings you cannot update:
- deliverPolicy
- filterSubject
- idleHeartbeat

Code Reference:
- [Server Implementation](https://github.com/nats-io/nats-server/blob/297b44394f92e990da450178cffaa7adecc88837/server/consumer.go#L1216)

Java unit tests:
- [testValidConsumerUpdates](https://github.com/nats-io/nats.java/blob/main/src/test/java/io/nats/client/impl/JetStreamManagementTests.java#L366)
- [testInvalidConsumerUpdates](https://github.com/nats-io/nats.java/blob/main/src/test/java/io/nats/client/impl/JetStreamManagementTests.java#L394)

In practice, this only applies to a durable consumer, but if you know the internal name of an ephemeral consumer, you might be able to update it's settings. This has not been verified.

**Comment by tbeets** [comment]
I assume `filterSubject` cannot be changed because doing so would change all prior sequencing and working set assumptions (i.e. fundamentally changing the view of the underlying stream).  Is that correct?

**Comment by ripienaar** [comment]
https://github.com/nats-io/nats-server/blob/297b44394f92e990da450178cffaa7adecc88837/server/consumer.go#L1216
